### PR TITLE
Dependency analysis

### PR DESCRIPTION
**Pull request by xeren**
Moves dependency code into a new class `...program.analysis.Dependency`, which before was distributed between `BasicRegRelation` and `EventCache`.  As a result, the relations `idd` and `addrDirect` are removed from `Wmm.BASE_RELATIONS`.
The analysis is more precise.  In the following program, `r := 0;` is no longer a may-dependency of `s := r;`.
```
r := 0;
if (b)
    r := 1;
else
    r := 2;
s := r;
```
<strike>The encoding now behaves different on uninitialized registers.  The following program now allows executions where `b` is false.  In this case, `s := r;` receives `r=0`.</strike>

Edit:  I misinterpreted the existing code.  Executions where `b` is false already do satisfy the encoding, with no further constraints on the value of `r`.
```
if (b)
    r := 1;
s := r;
```
Also, the interface of `Thread` is extended for this purpose.  I recommend making `Register(String,int,int)` package-private.

### CODE REVIEW

**Comment by ThomasHaas** [comment]
I think registers should not have init values. If they are uninitialized and an access to them can happen, then we have undefined behavior (UB)*. Compilers may still compile the code and in that case the content of a register is effectively non-deterministic during execution. So it is also reasonable to treat uninitialized registers as initialized with `non_det_X` where `X` is the type of the register. This also matches the SMT-solving in the sense that the register variable will be unconstrained and can take any value.


* There are cases where C allows the access to uninitialized locals without causing UB but only if the addresses of those locals are taken somewhere, but this forces the locals to be lifted to memory objects rather than registers by SMACK.

**Comment by hernanponcedeleon** [comment]
I haven't check the code in detail since the CI is still failing, however I like the idea of having a dependency analysis which might detect the non-dependency @xeren showed in the first example (I don't have a precise use case in mind, but this sounds like something we could use). That being said, I agree with @ThomasHaas that having the dependency analysis and the encoding all in a single class is not a good idea.

About uninitialised registers: I would expect the that for the second example, we have `b => r=1` and `!b => r can be whatever`. I could not see any new code which initialises registers and AFAIK we currently don't do that in `development` so I'm a little bit confused.

**Comment by ThomasHaas** [comment]
> About uninitialised registers: I would expect the that for the second example, we have `b => r=1` and `!b => r can be whatever`. I could not see any new code which initialises registers and AFAIK we currently don't do that in `development` so I'm a little bit confused.

I think the encoding handles this in this PR:
```
      if(uninitialized) {
                enc = bmgr.and(enc, bmgr.or(overwrite, bmgr.not(reader.cf()), generalZero(value, ctx)));
      }
```
That being said, I think the variable should be unconstrained to allow arbitrary values or we should throw an exception for malformed programs (since this is UB for C programs at least).

**Comment by xeren** [comment]
Regarding the handling of uninitialized registers, I must have misinterpreted `BasicRegRelation#doEncodeApprox(Collection,SolverContext)`.  I got confused with the following snippet:
```
if(writers.isEmpty() || writers.get(0).getCId() >= regReader.getCId()){
    BooleanFormula equal = generalEqual(register.toIntFormula(regReader, ctx),
        new IValue(BigInteger.ZERO, register.getPrecision()).toIntFormula(ctx), ctx);
    enc = bmgr.and(enc, equal);
} else {
```
This implies 0 as the expected value, if there are no writers to the register that are program-ordered before the reader.  This behavior is sensitive to preprocessing, as existing prior writers may i.e. be dead code and thus be eliminated from the program.
On the other hand, there are litmus tests that require this behavior, i.e. `litmus/C/auto/C-RW-G+RW-B+RW-G+RW-B+RW-G+RW-G+RW-r+RW-C.litmus`.  Thus, I have made the initialization optional.

**Comment by xeren** [comment]
In order to enforce my assumption, that after `ProcessingManager#run(Program)`, the program would be, besides others, free of dead code, I inserted a second instance of `DeadCodeElimination` into its pipeline.  Now it turns out that i.e. in the test `src/test/resources/locks/linuxrwlock-3.bpl`, this eliminates an assertion.  The failure happens, because `AssertInline` still refers to that event, but the event never got its encoding initialized.  I don't understand, how `SymmetryReduction` manages to remove assertions without this happening.

**Comment by ThomasHaas** [comment]
The issue with removing assertions is known and caused problems in some other PR (the one regarding forwarding of jumps I believe). IIRC, the bottom line is that when assertions get deleted, the assertions stored in the `Program` have to get updated. But to do this update, one has to know if the program is a litmus test or not, cause blindly recollecting program assertions fails for litmus tests as those define their assertions differently.
This is unfortunate and needs a rework. I think program assertions should get collected on demand and not stored so that deletions/additions are always visible without causing any inconsistent views. Assertions coming from litmus tests should get stored separately as those cannot get collected from the `Program` as they are not visible in the internal representation.

Also, I don't know why `SymmetryReduction` did not cause this problem but I wondered the same quite some time ago. I don't rememember if I found out though :).
We have it disabled by default, so maybe it would actually trigger the problem.

**Comment by hernanponcedeleon** [comment]
Clearly there is a read to the uninitialised register `r4` in [litmus/C/auto/C-RW-G+RW-B+RW-G+RW-B+RW-G+RW-G+RW-r+RW-C.litmus](https://github.com/hernanponcedeleon/Dat3M/blob/master/litmus/C/auto/C-RW-G%2BRW-B%2BRW-G%2BRW-B%2BRW-G%2BRW-G%2BRW-r%2BRW-C.litmus).

Apparently `herd7` initialises register to 0 by default
```
❯ cat test.litmus
X86 Test
""
{}
 P0;
 MOV [x],$1;
forall 0:EAX=0
❯ herd7 -model cat/tso.cat test.litmus
Test Test Required
States 1
0:EAX=0;
Ok
Witnesses
Positive: 1 Negative: 0
Condition forall (0:EAX=0)
Observation Test Always 1 0
Time Test 0.00
Hash=8bbea5a140818831c62a4e0a173102e0
```
For compatibility reason the best might be to initialise such registers to zero, but log a warning about UB.

**Comment by ThomasHaas** [comment]
Ok, so we have different semantics for `Litmus` and `C` code. I guess the code should be aware of what we are dealing with.
It might be reasonable to add an enum `{Litmus, C}` and store an instance in `Program`. Registers might want to have a default value field that can get set to some value or to `uninit` (maybe represented by NULL). Then you can have some preprocessing analysis that check for uninit regs and set their value to `0` for litmus code or otherwise throw an exception for `C` code.

**Comment by xeren** [comment]
As of now, it does not have to be that fine-grained.  This PR now adds an option `encoding.initializeRegisters`, which defaults to false and if enabled, encodes only those executions where an initial register value is zero.  Surprisingly, all checks pass without me setting it for litmus tests.  I could add one litmus test just to enforce this behavior for this project.

Interestingly, the recent failures originate from inter-thread register usage.  Given that a function `f` accepts one parameter `x`, `pthread_create(handle,info,&f,arg)` is parsed into an instance of `Create` and a new `Thread` starting with a `Start`, followed by a `Local` `x <- arg` and the body of `f`.  `Dependency` now explicitly accounts for that.  I imitated the existing behavior from `BasicRegRelation`, which effectively fetches the last change to the register by terms of `cId`.  Thanks to the usage of scopes, the communicated register is never overwritten, and thus draws an external `idd` or `addrDirect` edge.  I did not expect this.
A related issue features Boogie's global constants, which are implemented as registers of Thread `-1`, if not constrained to a single value.  If I interpret it correctly, it is treated in `development` like litmus-test registers:  There cannot be a writer, and thus the value has to be zero in all executions.

**Comment by ThomasHaas** [comment]
The case of passing function parameters for `pthread_create` via registers of the calling thread is indeed a peculiar case that is easy to miss/forget.

I'm not too sure about Boogie's global constants. Do we use them anywhere? I mean, we do inline all constants as far as I can tell.

**Comment by hernanponcedeleon** [comment]
Yes, I think at some point I was getting problems in constant propagation due to the use of registers to pass thread arguments. However the only other option I see would be to pass the arguments over memory, but that would increase relations like RF and thus deteriorate performance.

I don't really get the point about boogie global constants.

I'm also surprised that no test fails when the option is set to false. This suggests that the option is indeed not needed, but there was this litmus tests from LKMM which was failing due to reading an uninitialized Register.

**Comment by ThomasHaas** [comment]
I think using `rf` to pass arguments could work efficiently in practice, because you can fixate that rf-edge properly.
However, that seems to be unnecessary work. I think we could also allow special registers for that purpose, possibly tagged as `external` or `parameter` or something like that to distinguish those from true thread-local registers.
Maybe it would even make sense to add a different class `ExternalParam` (or `ExternalRegister extends Register`) that can reference a register (+ event where it got assigned) from a different thread. I think we might even be able to skip the encoding of `idd` for these altogether.

**Comment by xeren** [comment]
Currently, each instance of `INonDet` identifies a global variable, regardless of the event that evaluates the expression.  Together with `Assume`, this can implement inter-thread communication without memory events or dependency relationships.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorPower.java
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorPower.java:85]: This is unused

File: dartagnan/src/test/java/com/dat3m/dartagnan/miscellaneous/AliasAnalysisTest.java
Comment (hernanponcedeleon) [inline dartagnan/src/test/java/com/dat3m/dartagnan/miscellaneous/AliasAnalysisTest.java:44]: Why are these tests in the "AliasAnalysis" test class?

File: dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java
Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java:225]: Shortcut, since the whole entry set is available.  Both expressions are equivalent.

File: dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java
Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java:285]: Good idea.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java
Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java:57]: A dead register reader could have an empty list of writers, violating some assertions in `Dependency`.  A dead register writer could leave its reference behind, which would require more cleanup.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorPower.java
Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorPower.java:85]: Got it, thanks.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/Dependency.java
Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/Dependency.java:25]: In the following program, both dependency relationships (`r := 1`,`s := r`) and (`r := 2`,`s := r`) become active, when both participating events are executed.  Therefore, each reader can have an arbitrary number of must-writers.
```
if (b)
    r := 1;
else
    r := 2;
s := r;
```

File: dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java
Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java:69]: ok.

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/utils/ProgramBuilder.java
Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/utils/ProgramBuilder.java:149]: This was moved from `Register(String,int,int)`.  In order to get a reliable `Thread#getRegisters()`, I delegated nearly all usages of the constructor.  While I was on it, I moved the null-check and `dummyCount`, as well.
I am unsure whether `ProgramBuilder#getOrCreateRegister(int,String,int)` is passed `null`.

File: dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java:204]: `and` is reflexive, thus this conditional is useless

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/Dependency.java
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/Dependency.java:23]: The key in `map` and the list elements in `may` and `must` are always `RegReader` and `RegWriter`. Why, instead of using `Event`, you don't use the more precise type? This allows to simplify parts of the code like the assertion [here](https://github.com/hernanponcedeleon/Dat3M/blob/961a16415ffbec4b26882c6830c20ae36ef42ef0/dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java#L249) or the cast [here](https://github.com/hernanponcedeleon/Dat3M/blob/961a16415ffbec4b26882c6830c20ae36ef42ef0/dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java#L258) (there are probably more).

File: dartagnan/src/main/java/com/dat3m/dartagnan/expression/utils/Utils.java
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/expression/utils/Utils.java:25]: `generalEqualZero` is probably a better name

File: dartagnan/src/test/java/com/dat3m/dartagnan/miscellaneous/AliasAnalysisTest.java
Comment (hernanponcedeleon) [reply dartagnan/src/test/java/com/dat3m/dartagnan/miscellaneous/AliasAnalysisTest.java:44]: `AnalysisTest` sounds better

**Review by hernanponcedeleon — CHANGES_REQUESTED** [review]

File: dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java:204]: I think you mean commutative. That is in principle true, but we actually need to check if that makes a difference in performance in any way, because it can simplify detection of identical subterms.

I just noticed, that you took the code from `Relation.getExecVar` and adapted it.
I think more maximal sharing, you still might want to order `x` and `y` by `cId` first.
There is this hypothetical case where both `x` and `y` imply each other but are not merged (i.e. their `exec`-variables are different). This case shouldn't happen as long as we merge `exec`-vars though.

File: dartagnan/src/main/java/com/dat3m/dartagnan/expression/utils/Utils.java
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/expression/utils/Utils.java:25]: I agree.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java:57]: If possible, I would keep the preconditions as few as possible. If there is a way to treat dead reads/writes without much issues, I think that is a better solution. If it is hard to do and the DCE is really needed, we should definitely put some comment on the class that lists this requirement (maybe you already did, I didn't check).

File: dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java:204]: Yes, I meant commutative.

I don't get you following comment. Adding `A /\ B` or `B /\ A` to the solver should be exactly the same. The constraints are equivalent and the solver should decide internally how to rewrite this in whatever way it is more efficient for it. Deciding the order of terms should not be the responsibility of the user. Otherwise we will have thousend of decisions all over the code.

Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java:204]: I agree that **simplification** does improve solver performance, but here we are talking about equivalent constraints with the same number of sub-terms (more a reordering rewriting than a simplification). If it is really true that the solver performance would be better if we add `A /\ B` instead of `B /\ A`, then where to draw the line? Shall we add `programEncoder /\ wmmEncoder` or `wmmEncoder /\ programEncoder` ... this looks like a rabbit hole to me.

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java:204]: I think you misunderstood the intention here. It is not about `A /\ B` vs. `B /\ A` but it is about deciding on one of them and sticking to it throughout the whole formula. If you don't, then you find both variations in the same formula and depending on the solver, he might not be able to identify both expressions as equivalent (since they are syntactically distinct). 
The general question of which order is the better one is not answerable, it heavily depends on the actual solver and the actual instance you look at. But I think it is generally true that it is better to avoid asserting different permutations of the same formula because the solver might fail to identify them as equivalent.

**Comment by ThomasHaas** [comment]
I took a look into the encoding of the local data flow and there is heavy usage of `reader.cf` instead of `reader.exec`. I think this code is already in `development/master` like this, but does anybody see why?
Is it perhaps related to the fact that some conditional events need to take their value from somewhere even if they end up not actually executing (since, e.g., the value fails the condition)? If so, then care has to be taken when deciding on `cf` vs. `exec`.
I think the only events that are affected by this difference are the Linux `RMWCond` events which can read values without actually executing.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/Dependency.java
Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/Dependency.java:23]: There is no `RegReader` interface, but `RegReaderData` and `MemEvent`, and I see no benefit in introducing such a new interface.
`RegWriter` does not extend the interface of `Event`, and thus had to be cast to `Event` when needed, like `writer.exec()` or `new Tuple(writer,reader)`.  This was one of the few points of #202.

**Comment by hernanponcedeleon** [comment]
LGTM so unless @ThomasHaas has further comments, I'll merge.

BTW: I saw one of the last commits set the option for litmus test to initialise registers. I would have expected this to be necessary, but before to this commit it seems it was not really needed. Why was that?

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorArm8.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorArm8.java:108]: This register was named before and I think we should keep that name.

**Review by ThomasHaas — CHANGES_REQUESTED** [review]

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java
Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java:57]: This line was a temporary fix for RemoveDeadCondJumps introducing new dead code.  As mentioned in #219, DCE is meant to be run before unrolling.  Therefore, i'll add a small DCE to `RemoveDeadCondJumps`, such that it cleans up after itself.

Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java:57]: There is something that it is not clear to me yet (the first comment by @xeren was supposed to answer this, but I did not really get it) ... do we really need the first DCE pass? Can we just keep the second DCE pass (the one before unrolling) so it takes care of cleaning any dead code another pass may introduce?

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java:57]: If the DCE before `BranchReordering` is removed, branch reordering will order unreachable code to before `Thread.getEntry` resulting in pseudo-deleted code. It is deleted in the sense that `Thread.getEvents` won't return that code anymore because it is unreachable starting from `Thread.getEntry`. However, it does not get garbage collected because we keep (unintended) references to the dead code via e.g. `Label.getListeners`. 
Overall, this can cause memory leaks. A subsequent run of `DCE` cannot delete that code because it is unreachable via `Event.getSuccessor`.
In short: the first `DCE` is crucial as long as we use `BranchReordering`

Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java]: This change should have been undone.  Instead, `RemoveDeadCondJumps` now performs its own DCE.

Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java]: You mean, `Dependency` should not assert that certain processes like DCE were performed beforehand, such that they still can be opted out when desired.  There is still one assertion that has to stay for now: That the thread that calls `pthread_create` appears before the created thread in the list of threads.

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java]: Yes, I think it is best if the preconditions under which an analysis or a program transformation runs should be as small as possible. If handling dead code requires a couple of lines of extra code but removes the precondition of having all dead code eliminated, I would rather add the few extra lines of code to buy flexibility. And if the analysis just becomes inaccurate without proper preprocessing then so be it. In that case it is our fault for not doing the preprocessing beforehand.
If the analysis is simply not possible without specific preprocessing steps, then that is fine and we can require those steps beforehand.
I believe some of the existing analysis/preprocessors have unnecessary requirements (like DCE being done before unrolling) and we should eventually remove those as well.

Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java]: Good point.  After I have added some logging, I think this PR will be ready to be merged.

**Comment by hernanponcedeleon** [comment]
To follow the same pattern we have for the other analysis, it would be better to use a factory method (`fromConfig`) rather than `new Dependency`. If you pass the `analysisContext` to `fromConfig`, you don't need to pass the `ExecutionAnalysis` to the constructor (you can obtain it via `requires` similarly to how `ExecutionAnalysis` obtains its `BranchEquivalence`). This should simplify a little bit the code in `VerificationTask`.

With this final change I think this PR could be merged.

**Pull request merged**
